nix/modules/tezos-node.nix

```diff
@@ -0,0 +1,134 @@
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  tezos-node-pkg = pkgs.tezos.binaries.tezos-node;
+  cfg = config.services.tezos-node;
+  genConfigCommand = historyMode: rpcPort: netPort: network: ''
+                      --data-dir "$node_data_dir" \
+                      --history-mode "${cfg.historyMode}" \
+                      --rpc-addr ":${toString cfg.rpcPort}" \
+                      --net-addr ":${toString cfg.netPort}" \
+                      --network "${cfg.network}"
+  '';
+in {
+  options.services.tezos-node = rec {
+    enable = mkEnableOption "Tezos node service";
+
+    node-name = mkOption {
+      type = types.str;
+      default = "edonet";
+      description = ''
+        Local name of the node, to support running multiple local instances.
+      '';
+    };
+
+    package = mkOption {
+      default = tezos-node-pkg;
+      type = types.package;
+    };
+
+    rpcPort = mkOption {
+      type = types.int;
+      default = 8732;
+      example = 8732;
+      description = ''
+        Tezos node RPC port.
+      '';
+    };
+
+    netPort = mkOption {
+      type = types.int;
+      default = 9732;
+      example = 9732;
+      description = ''
+        Tezos node net port.
+      '';
+    };
+
+    logVerbosity = mkOption {
+      type = types.str;
+      default = "warning";
+      description = ''
+        Level of logs verbosity. Possible values are:
+        fatal, error, warn, notice, info or debug.
+      '';
+    };
+
+    network = mkOption {
+      type = types.str;
+      default = "edonet";
+      description = ''
+        Network which node will be running on.
+      '';
+    };
+
+    historyMode = mkOption {
+      type = types.str;
+      default = "full";
+      description = ''
+        Node history mode. Possible values are:
+        full, experimental-rolling or arcive.
+      '';
+    };
+
+    nodeConfig = mkOption {
+      default = null;
+      type = types.nullOr pkgs.serokell-nix.lib.types.jsonConfig;
+      description = ''
+        Custom node config.
+        This option overrides the all other options that affect
+        tezos-node config.
+      '';
+    };
+  };
+  config = mkIf cfg.enable {
+    users.groups."tezos-${cfg.node-name}" = { };
+    users.users."tezos-${cfg.node-name}" = { group = "tezos-${cfg.node-name}"; };
+    systemd.services."tezos-${cfg.node-name}-tezos-node" = rec {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      description = "Tezos node";
+      environment = {
+        TEZOS_LOG = "* -> ${cfg.logVerbosity}";
+      };
+      serviceConfig = {
+        User = "tezos-${cfg.node-name}";
+        StateDirectory = "tezos-${cfg.node-name}";
+        Restart = "always";
+        RestartSec = "10";
+      };
+      preStart =
```

Comment:
Not sure why so much copy-paste in this option. You can do things like:

```nix
{

x = ''
    hello world
    blabla
  '' + lib.optionalString (cfg.nodeConfig != null) ''
    foo bar
  '';

}
```

Also, this whole custom `nodeConfig` story is slightly suspicious. The standard pattern is to use the module options to generate a config file, and then append something like `extraOptions` to the end. (In the case of JSON you can just make `extraOptions` an attrset, then generate an attrset from options, and finally merge the two attrsets, and then use `builtins.toJSON` to write your config file.)

Reply:
I'm not sure if what you suggest if a great idea: `tezos-node` does some magical config construction that I don't really want to touch directly for fear of overwriting something important. E.g. look at the default config:

```
{ "p2p":
    { "bootstrap-peers":
        [ "boot.tzbeta.net", "116.202.172.21", "95.216.45.62" ],
      "listen-addr": "[::]:9732" } }
```

Updating it through cli while leaving the option for overwriting with a full ready-made config seems a bit more reasonable. Am I wrong?